server/homebrew.api.js

```diff
@@ -89,148 +129,177 @@ const newBrew = async (req, res)=>{
 
 	beforeNewSave(req.account, brew);
 
-	let saved;
+	const newHomebrew = new HomebrewModel(brew);
+	newHomebrew.editId = nanoid(12);
+	newHomebrew.shareId = nanoid(12);
+
+	let googleId, saved;
 	if(transferToGoogle) {
-		saved = await newGoogleBrew(req.account, brew, res)
+		googleId = await newGoogleBrew(req.account, newHomebrew, res)
 			.catch((err)=>{
-				res.status(err.status || err.response.status).send(err.message || err);
+				res.status(err?.status || err?.response?.status || 500).send(err?.message || err);
 			});
+		excludeStubProps(newHomebrew);
 	} else {
-		saved = await newLocalBrew(brew)
-			.catch((err)=>{
-				res.status(500).send(err);
-			});
+		// Compress brew text to binary before saving
+		newHomebrew.textBin = zlib.deflateRawSync(newHomebrew.text);
+		// Delete the non-binary text field since it's not needed anymore
+		newHomebrew.text = undefined;
+	}
+	if(transferToGoogle && !googleId) {
+		if(!res.headersSent) {
+			res.status(500).send('Unable to save document to Google Drive');
+		}
+		return;
+	} else if(transferToGoogle) {
+		newHomebrew.googleId = googleId;
 	}
+
+	saved = await newHomebrew.save()
+		.catch((err)=>{
+			console.error(err, err.toString(), err.stack);
+			throw `Error while creating new brew, ${err.toString()}`;
+		});
 	if(!saved) return;
+	saved = saved.toObject();
+
 	return res.status(200).send(saved);
 };
 
 const updateBrew = async (req, res)=>{
-	let brew = excludePropsFromUpdate(req.body);
+	console.log(req.brew, req.body);
+	let brew = Object.assign(req.brew, excludePropsFromUpdate(req.body));
+	brew.text = mergeBrewText(brew);
 	const { transferToGoogle, transferFromGoogle } = req.query;
 
-	let saved;
 	if(brew.googleId && transferFromGoogle) {
-		beforeNewSave(req.account, brew);
-
-		saved = await newLocalBrew(brew)
+		const deleted = await deleteGoogleBrew(req.account, brew.googleId, brew.editId, res)
 			.catch((err)=>{
 				console.error(err);
-				res.status(500).send(err);
+				res.status(err?.status || err?.response?.status || 500).send(err.message || err);
 			});
-		if(!saved) return;
+		if(!deleted) {
+			if(res.headersSent) {
+				res.status(500).send('Unable to delete brew from Google');
```

Comment:
Somehow this seems redundant with the line above where we are already sending an error in the `catch`. In fact, should we move all the error reporting from here into `deleteGoogleBrew` so any call to `deleteGoogleBrew` will automatically have errors handled?

Keeping the `if` is fine so we can `return`.

Reply:
This in particular is a good example of why I prefer try/catch instead of `.catch`. If we used try/catch in these cases and didn't have the `.catch` then at the end of the catch we could return.

I'm not really fond of passing the response around between files, partially because it makes understanding where things are happening more difficult. 

Considering the repetitive nature of this particular if statement, we could write a middleware piece that sends a 500 if no headers have been sent that runs after all other middlewares.

Reply:
At the very bottom of `api.js` we already have an error-handling middleware that does this for any uncaught errors. Is that what you had in mind?

Reply:
_Partially_ that, but it's more a "I'm the last middleware and no headers have been sent back to the client so I'm going to send something so it doesn't time out" instead of sending a particular error passed into the middleware

Reply:
Do you think the existing Middleware can be updated to handle errors in the way you want?

Reply:
The existing middleware expects an error as the first piece passed in, whereas the middleware I'm intending to write would not. It would simply handle any cases where the backend failed to send headers for some reason. While that _shouldn't_ ever happen, if it did we would end up with request timeouts.